component/client-handler/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/jwt/validator/JWTValidator.java

```diff
@@ -72,185 +73,322 @@
+                issuedTime = issuedAtTime.getTime();
+            }
+
+            //Validate mandotory feilds, audience, nbf,exp time, jti.
+            if (!validateMandatoryFeilds(mandatoryClaims, claimsSet) || !validateAudience(validAud, audience) ||
+                    !validateJWTWithExpTime(expirationTime, currentTimeInMillis, timeStampSkewMillis) ||
+                    !validateNotBeforeClaim(currentTimeInMillis, timeStampSkewMillis, nbf) ||
+                    !validateJTI(signedJWT, jti, currentTimeInMillis, timeStampSkewMillis, expTime, issuedTime) ||
+                    !validateAgeOfTheToken(issuedAtTime, currentTimeInMillis, timeStampSkewMillis)) {
+                return false;
+            }
+
+            return true;
+
+        } catch (IdentityOAuth2Exception e) {
+            throw new OAuthClientAuthnException(e.getMessage(), OAuth2ErrorCodes.INVALID_REQUEST);
         }
+    }
 
-        if (!validateJTI(signedJWT, jti, currentTimeInMillis, timeStampSkewMillis, expirationTime.getTime(),
-                issuedAtTime.getTime())) {
-            return false;
+    private boolean validateMandatoryFeilds(List<String> mandatoryClaims, ReadOnlyJWTClaimsSet claimsSet) throws OAuthClientAuthnException {
+
+        for (String mandatoryClaim : mandatoryClaims) {
+            if (claimsSet.getAllClaims().get(mandatoryClaim) == null) {
+                String errorMessage = "Mandatory field/feilds (Issuer, Subject, Expiration time , JWT ID or " +
+                        "Audience) are missing in the JWT assertion";
+                if (log.isDebugEnabled()) {
+                    log.debug(errorMessage);
+                }
+                throw new OAuthClientAuthnException(errorMessage, OAuth2ErrorCodes.INVALID_REQUEST);
+            }
         }
+        return true;
+    }
+
+    //"REQUIRED. sub. This MUST contain the client_id of the OAuth Client."
+    public boolean validateSubject(String jwtSubject) throws OAuthClientAuthnException {
 
-        //validate whether the subject is client_id
         OAuthAppDO oAuthAppDO = getOAuthAppDO(jwtSubject);
-        if (oAuthAppDO == null) {
-            return logAndReturnFalse("Unable to find OAuth application with provided JWT information with subject:"
-                    + jwtSubject);
-        }
-        //check whether the issuer is client_id
-        if (isEmpty(validIssuer) && !jwtIssuer.equals(oAuthAppDO.getOauthConsumerKey())) {
-            return logAndReturnFalse("Invalid field Issuer:" + jwtIssuer + " in the given JSON Web Token.");
-        }
+        return validateWithClientId(jwtSubject, oAuthAppDO);
+    }
 
-        String tenantDomain = oAuthAppDO.getUser().getTenantDomain();
-        //validate signature
-        if (!isValidSignature(signedJWT, tenantDomain, jwtSubject)) {
-            return logAndReturnFalse("Signature or Message Authentication invalid for:" + jwtSubject);
+    //"REQUIRED. iss. This MUST contain the client_id of the OAuth Client."
```

Comment:
This comment is not always true. only If the valid issuer is specified in the jwtValidator, then issuer should be validated against client_id.